app/scripts/controllers/network-order.ts

```diff
@@ -109,51 +108,37 @@ export class NetworkOrderController extends BaseController<
    *
    * @param networkControllerState - The state of the network controller.
+   * @param networkControllerState.networkConfigurationsByChainId
    */
-  onNetworkControllerStateChange(networkControllerState: NetworkState) {
-    // Extract network configurations from the state
-    const networkConfigurations = Object.values(
-      networkControllerState.networkConfigurations,
-    );
-
-    // Since networkConfigurations doesn't have default or mainnet network configurations we need to combine mainnet chains with network configurations
-    const combinedNetworks = [...MAINNET_CHAINS, ...networkConfigurations];
-
-    // Extract unique chainIds from the combined networks
-    const uniqueChainIds = combinedNetworks.map((item) => ({
-      networkId: item.chainId,
-      networkRpcUrl: item.rpcUrl,
-    }));
-
-    // Arrays to store reordered and new unique chainIds
-    let reorderedNetworks: NetworksInfo[] = [];
-    const newUniqueNetworks: NetworksInfo[] = [];
-
-    // Iterate through uniqueChainIds to reorder existing elements
-    uniqueChainIds.forEach((newItem) => {
-      const existingIndex = this.state.orderedNetworkList.findIndex(
-        (item) =>
-          item.networkId === newItem.networkId &&
-          item.networkRpcUrl === newItem.networkRpcUrl,
-      );
-      // eslint-disable-next-line no-negated-condition
-      if (existingIndex !== -1) {
-        // Reorder existing element
-        reorderedNetworks[existingIndex] = newItem;
-      } else {
-        // Add new unique element
-        newUniqueNetworks.push(newItem);
-      }
-    });
-
-    // Filter out null values and concatenate reordered and new unique networks
-    reorderedNetworks = reorderedNetworks
-      .filter((item) => Boolean(item))
-      .concat(newUniqueNetworks);
-
-    // Update the state with the new networks list
+  onNetworkControllerStateChange({
+    networkConfigurationsByChainId,
+  }: NetworkState) {
     this.update((state) => {
-      state.orderedNetworkList = reorderedNetworks;
+      // Filter out testnets, which are in the state but not orderable
+      const chainIds = Object.keys(networkConfigurationsByChainId).filter(
+        (chainId) =>
+          ![
+            CHAIN_IDS.SEPOLIA,
+            CHAIN_IDS.LINEA_SEPOLIA,
+            CHAIN_IDS.GOERLI,
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          ].includes(chainId as any),
+      ) as Hex[];
+
+      state.orderedNetworkList = state.orderedNetworkList
+        // Filter out deleted networks
+        .filter(({ networkId }) => chainIds.includes(networkId))
+        // Append new networks to the end
+        .concat(
+          chainIds
+            .filter(
+              (chainId) =>
+                !state.orderedNetworkList.some(
+                  ({ networkId }) => networkId === chainId,
+                ),
+            )
+            .map((chainId) => ({ networkId: chainId })),
```

Comment:
nit. thoughts on assigning this value to a variable instead of inlining for readability?

Reply:
https://github.com/MetaMask/metamask-extension/pull/26433/commits/15ac698e44ceba7dba661a2c08d9fffff3b3648c